maintenance: add branch management guide (#249)
inveniosoftware/docs-invenio-rdm: docs/maintenance/branch-management.md

# Branch management

**Intended audience**

This guide is intended for maintainers and developers of InvenioRDM itself.

**Scope**

The guide covers how branches are managed in InvenioRDM in order to support
the maintenance policy.

## Overview

An InvenioRDM Long-Term Support (LTS) release is supported for minimum 1 year
and for minimum 6 months after the next new LTS release (see
[maintenance policy](../releases/maintenance-policy.md)). Only the latest minor
version of an LTS release is supported (e.g. we do not support both v6.0 and

> v6.1).

Reviewer: :100:

### Opening pull requests

#### PR against ``master``

You should normally always open pull requests against the ``master`` branch.
One exception is when a bug/security fix is only reproducible on a maintenance

> branch.

Reviewer: good point

#### Separate logical commits

If you are fixing a bug, please ensure that the bug is fixed in a separate

> logical commit, to ensure that it easy to backport the the fix to previous
> maintained releases.

Reviewer: ```suggestion
logical commit, to ensure that it is easy to backport the fix to previous
maintained releases.
```

> #### Security fixes is done in hidden branches
> 
> Note that security fixes is normally done in a coordinated way using hidden
> branches until PyPI releases have been issued. Please get in touch with a
> maintainer.

Reviewer: ```suggestion
#### Security fixes are done in hidden branches

Note that security fixes are normally done in a coordinated way using hidden
branches until PyPI releases have been issued. Please get in touch with a
maintainer.
```

other good scenario to outline :+1:

### Merging pull requests

> A PR is merged as usual once the continues integration tests passes and the
> code has been through a code review.

Reviewer: ```suggestion
A PR is merged as usual once the continuous integration tests pass and the
code has been through a code review.
```

#### Maintainer is responsible for backporting

The maintainer merging the PR is now responsible for backporting the fix to all
maintained releases. The maintainer is free to ask the developer opening the
PR to backport the fix, but it's the maintainer that's ultimately responsible
for ensuring that the fix is backported.

#### Supported maintenance branches

> Following is an overview which versions/branch(es) are currently supported for

Reviewer: ```suggestion
Following is an overview of which versions/branch(es) are currently supported for
```

| Module                      | Supported versions | Supported branches |
| --------------------------- | ------------------ | ------------------ |
| flask-resources             | ``0.7.x``          | ``maint-0.7``      |
| invenio                     | ``3.5.0a2``        | ``maint-3.5``      |
| invenio-app-rdm             | ``6.0.x``          | ``maint-6.0``      |
| invenio-communities         | ``2.5.x``          | ``maint-2.5``      |
| invenio-drafts-resources    | ``0.13.x``         | ``maint-0.13``     |
| invenio-rdm-records         | ``0.32.x``         | ``maint-0.32``     |
| invenio-records-permissions | ``0.12.x``         | ``maint-0.12``     |
| invenio-records-resources   | ``0.16.x``         | ``maint-0.16``     |
| invenio-s3                  | ``1.0.x``          | ``maint-1.0``      |
| invenio-vocabularies        | ``0.8.x``          | ``maint-0.8``      |
| marshmallow-utils           | ``0.5.x``          | ``maint-0.5``      |
| react-invenio-deposit       | ``0.16.x``         | ``maint-0.16``     |
| react-invenio-forms         | ``0.8.x``          | ``maint-0.8``      |

!!! info "Didn't find the module?"

>     If a module is not listed below it is because the version is either
>     unconstrained or part of Invenio Framework and constrained by the Invenio
>     Framework setup. For unconstrained modules you should take create care that
>     the change/fix is backward compatible. If not, we have to constrain the
>     module version in one of the InvenioRDM modules.

Reviewer: ```suggestion
    If a module is not listed above, it is because the version is either
    unconstrained or part of Invenio Framework and constrained by the Invenio
    Framework setup. For unconstrained modules you should take great care that
    the change/fix is backward compatible. If not, we have to constrain the
    module version in one of the InvenioRDM modules.
```
